src/ralph/admin/fields.py

```diff
@@ -0,0 +1,147 @@
+        non_empty_values = [item for item in values if len(item.strip()) > 0]
+        if not self.allow_duplicates:
+            has_duplicates = len(set(non_empty_values)) != len(non_empty_values)
+            if has_duplicates:
+                raise forms.ValidationError(_("There are duplicates in field."))
+
+    def to_python(self, value):
+        items = []
+        if value:
+            for item in re.split(self.separators, value):
+                items.append(item.strip(' \t\n\r'))
+
+        return items
+
+
+class MultivalueFormMixin(object):
+    """A form that has several multiline fields that need to have the
+    same number of entries.
+
+    :param multivalue_fields: list of form fields which require the same item
+    count
+    :param one_of_mulitvalue_required: list of form fields names which must
+    be provided to form, e.g.
+        one_of_mulitvalue_required = ['sn', 'barcode']
+        means that if user inputs such data:
+        sn, barcode:
+        1,1, # valid
+        2,   # valid
+        ,2   # valid
+        ,    # invalid, because `sn` neither `barcode` was provided
+    """
+    multivalue_fields = []
+    one_of_mulitvalue_required = []
+    model = None
+
+    def equal_count_validator(self, cleaned_data):
+        """Adds a validation error if form's multivalues fields have
+        different count of items."""
+        items_count_per_multi = set()
+        for field in self.multivalue_fields:
+            if cleaned_data.get(field, []):
+                items_count_per_multi.add(len(cleaned_data.get(field, [])))
+        if len(items_count_per_multi) > 1:
+            for field in self.multivalue_fields:
+                if field in cleaned_data:
+                    msg = _("Fields: %(fields)s - require the same count") % {
+                        'fields': ', '.join(self.multivalue_fields)
+                    }
+                    self.errors.setdefault(field, []).append(msg)
+
+    def any_in_multivalues_validator(self, data):
+        def rows_of_required():
+            rows_of_required = [
+                self.cleaned_data[field_name] for field_name in
+                self.one_of_mulitvalue_required if field_name in
+                self.cleaned_data
+            ]
+            for multivalues_row in zip(*rows_of_required):
+                yield multivalues_row
+
+        if self.one_of_mulitvalue_required:
+            for row_of_required in rows_of_required():
+                if not any(row_of_required):
+                    for field_name in self.one_of_mulitvalue_required:
+                        errors = self._errors.setdefault(
+                            field_name, ErrorList()
+                        )
+                        msg = _('Fill at least on of %(v)s in each row') % {
+                            'v': ','.join(self.one_of_mulitvalue_required)
+                        }
+                        errors.append(_(msg))
```

Comment:
I think the current version is better, because error message is showed under each field which validation is applied to.